### ISSUE TITLE
Issue when adding contacts to tradingAccountEntry

### ISSUE THREAD

**Issue by CodeBetaIIndia** [Sage/SData-2.0#78]
Hi, I have an issue with adding a customer with contact. When I am trying to add a contact to the tradingAccount object it shows an error "unsupported media type". Below is my code snippet.

            //phoneNumberFeed phoneNumberFeed = new phoneNumberFeed();
            //phoneNumberFeed.Entries.Add(new phoneNumberFeedEntry
            //{
            //    number = "+61 (0)2 0000 0000"
            //});
            //emailFeed emailFeed = new emailFeed();
            //emailFeed.Entries.Add(new emailFeedEntry
            //{
            //    address = "test@testmail.com"
            //});
            //contactFeed contactsFeed = new contactFeed();
            //contactsFeed.Entries.Add(new contactFeedEntry
            //{
            //    firstName = "FName",
            //    fullName = "Fname LName",
            //    phones = phoneNumberFeed,
            //    emails = emailFeed                
            //});
Is there any documentation or something for SIF?

**Comment by darroncockram** [comment]
The error message suggests you are not POSTing data that adheres to the Sdata standard. It would to be valid XML and valid as per the contract the adapter is using. 

The code snippet just shows creating the in-memory data rather than how you are submitting the data. If you could show the full code or the resulting request data I could probably advise further. 

In terms of documentation on the SIF specifically there is not a great deal. Best place to start would be this series of videos: www.youtube.com/playlist?list=PLF56D8511E5ABB680&feature=plcp.

**Comment by CodeBetaIIndia** [comment]
Hi Darron,
                 Thank you for the response. I am posting the code here, please give some suggestions to improve it. 
-------------------------------------------------------------------------------------------------------
tradingAccountFeedEntry tradingAccountTest = new tradingAccountFeedEntry();
Sage.Common.Syndication.SDataUri tradingAccountTestUri = new Sage.Common.Syndication.SDataUri();
tradingAccountTestUri.BuildLocalPath("Accounts50", "GCRM", "-", "tradingAccounts");

tradingAccountTest.customerSupplierFlag = "Customer";
tradingAccountTest.name = "Ciompany Name";

phoneNumberFeed phoneNumberFeed = new phoneNumberFeed();
phoneNumberFeed.Entries.Add(new phoneNumberFeedEntry
{
    number = "+61 (0)2 9266 5331"
});
emailFeed emailFeed = new emailFeed();
emailFeed.Entries.Add(new emailFeedEntry
{
    address = "testContact@testmail.com.au"
 });
 contactFeed contactsFeed = new contactFeed();
 contactsFeed.Entries.Add(new contactFeedEntry
 {
      firstName = "fcname",
      fullName = "fname",
      phones = phoneNumberFeed,
      emails = emailFeed
});

tradingAccountTest.contacts = contactsFeed;

SDataRequest accountRequestCreateTest = new SDataRequest(tradingAccountTestUri.Uri, tradingAccountTest, Sage.Integration.Messaging.Model.RequestVerb.POST);
accountRequestCreateTest.Username = "MANAGER";
accountRequestCreateTest.Password = "";

 tradingAccountFeedEntry tradingAccountResponseTest = new tradingAccountFeedEntry();
var response = accountRequestCreateTest.RequestFeedEntry<tradingAccountFeedEntry>(tradingAccountResponseTest);

-------------------------------------------------------------------------------------------------------

**Comment by darroncockram** [comment]
My guess is you are using Sage 50 Accounts and the problem is that you are not telling it what type of contact you want to create. You'll find that the IsStatusValidForVerb proeprty is set to false and the Diagnosis property contains details of an error.

If you set the contact type the adapter should them be able to processes it: 

```c#
contactsFeed.Entries.Add(new contactFeedEntry
{
	type = "Customer Delivery Contact",
	fullName = "FirstName LastName"
});`

**Comment by CodeBetaIIndia** [comment]
Thank you Darron,
                               You are right, its working now with type as "Customer Delivery Contact". But I got the same error when creating it as "Customer Communication" type. Is that because of incorrect type name?.
Thanks in advance.

**Comment by darroncockram** [comment]
> I got the same error when creating it as "Customer Communication" type. Is that because of incorrect type name?

Yes and no. The type is not invalid in that it is a type that the Sage 50 Accounts adapter exposes, but it is invalid in the context you are trying to use it.

The Customer Communication type relates to contacts that exist as part of a communication record rather than 'addresses and contacts' related a customer record (all of that is Sage 50 Accounts terminology rather than Sdata or the GCRM contract). It really comes down to an internal implementation but you cannot create a communication contact as part of a tradingAccount as there is no direct relationship. A 'Communication Contact' is related to an interaction type so it would have to be created when creating an interaction type. Unfortunately the interaction feed in the Sage 50 Accounts adapter is read only (you can see this in the schema where the feed is marked as sme:CanPost="false").

So in short if you want to create a contact as part of a tradingAccount you will need to set it's type to "Customer Delivery Contact" or "Supplier Delivery Contact", depending on what type of tradingAccount record you are creating, as these would be the only supported types for that scenario.

Hope that helps.

**Comment by CodeBetaIIndia** [comment]
Thank you Darron. I understand the limitations of external tools to handle Sage 50 Accounts from your valuable suggestions and support. Do we have any other open source to handle all the functionalities in Sage 50 other that SIF that you can suggest. If we have any other frameworks please put it here otherwise leave it. 
Once again thank you very much Darron.

**Comment by darroncockram** [comment]
Sdata is the only open/free mechanism available. SDO, the product specific SDK for Sage 50 Accounts offers a lot more functionality and would allow you do almost anything with the Accounts data but that is not free  I'm afraid. If you do want to use SDO you'd need to join the developer programme. Here's the link in case it is of interest: http://www.sage.co.uk/become-a-sage-developer/sme-business-software-developer.
